Additional EFO xref context from axioms

bfoltyn opened the pull request:
#18

bfoltyn commented:
@dhimmel For now, I've only added the SPARQL queries:
-  `nxontology_data/efo/queries/mapping_properties.rq` - Retrieves the mapping properties from classes
- `nxontology_data/efo/queries/sources.rq` - Retrieves the sources for xrefs.

Could you please review these queries? Let me know if there are any necessary changes.
Once the queries are done, what would be the next steps? How this data should be represented in the output?

dhimmel commented on nxontology_data/efo/queries/mapping_properties.rq:
```diff
@@ -0,0 +1,28 @@
+  )
+  BIND( REPLACE( STR(?mapping_property_uri), "^http://purl\\.obolibrary\\.org/obo/mondo#(.+)$", "mondo:$1" ) AS ?mapping_property_id )
+  BIND( REPLACE( STR(?mapping_property_id), "^http://www\\.w3\\.org/2004/02/skos/core#(.+)$", "skos:$1" ) AS ?mapping_property_id )
+}
```
Can you reply to this comment with the head of the output table?

bfoltyn replied:
efo_id | xref_id | mapping_property_id | efo_uri | xref_uri | mapping_property_uri
-- | -- | -- | -- | -- | --
MONDO:0000044 | meddra:10060873 | mondo:closeMatch | http://purl.obolibrary.org/obo/MONDO_0000044 | http://identifiers.org/meddra/10060873 | http://purl.obolibrary.org/obo/mondo#closeMatch
MONDO:0000050 | meddra:10035083 | mondo:closeMatch | http://purl.obolibrary.org/obo/MONDO_0000050 | http://identifiers.org/meddra/10035083 | http://purl.obolibrary.org/obo/mondo#closeMatch
MONDO:0000088 | meddra:10044701 | mondo:closeMatch | http://purl.obolibrary.org/obo/MONDO_0000088 | http://identifiers.org/meddra/10044701 | http://purl.obolibrary.org/obo/mondo#closeMatch
MONDO:0000088 | meddra:10058084 | mondo:closeMatch | http://purl.obolibrary.org/obo/MONDO_0000088 | http://identifiers.org/meddra/10058084 | http://purl.obolibrary.org/obo/mondo#closeMatch
MONDO:0000127 | meddra:10063361 | mondo:closeMatch | http://purl.obolibrary.org/obo/MONDO_0000127 | http://identifiers.org/meddra/10063361 | http://purl.obolibrary.org/obo/mondo#closeMatch

dhimmel commented on nxontology_data/efo/queries/sources.rq:
```diff
@@ -0,0 +1,17 @@
+  BIND( REPLACE( STR(?source), "^http.+/([^:]+)_(.+)$", "$1:$2" ) AS ?efo_id )
+}
+
+GROUP BY ?efo_id ?xref ?axiom_source
```
Can you reply to this comment with the head of the output table?

bfoltyn replied:
efo_id | xref | axiom_source
-- | -- | --
CHEBI:100241 | Beilstein:3568352 | Beilstein
CHEBI:100241 | CAS:85721-33-1 | ChemIDplus
CHEBI:100241 | CAS:85721-33-1 | KEGG COMPOUND
CHEBI:100241 | Drug_Central:659 | DrugCentral
CHEBI:100241 | PMID:10397494 | ChEMBL

dhimmel commented on nxontology_data/efo/queries/sources.rq:
```diff
@@ -0,0 +1,17 @@
+  BIND( REPLACE( STR(?source), "^http.+/([^:]+)_(.+)$", "$1:$2" ) AS ?efo_id )
+}
+
+GROUP BY ?efo_id ?xref ?axiom_source
```
Might be nice to add ORDER BY here.

bfoltyn replied:
efo_id | xref | axiom_source
-- | -- | --
CHEBI:100241 | Beilstein:3568352 | Beilstein
CHEBI:100241 | CAS:85721-33-1 | ChemIDplus
CHEBI:100241 | CAS:85721-33-1 | KEGG COMPOUND
CHEBI:100241 | Drug_Central:659 | DrugCentral
CHEBI:100241 | PMID:10397494 | ChEMBL

bfoltyn replied:
Ok

dhimmel commented on nxontology_data/efo/queries/sources.rq:
```diff
@@ -0,0 +1,17 @@
+PREFIX oboInOwl: <http://www.geneontology.org/formats/oboInOwl#>
+
+SELECT ?efo_id ?xref  ?axiom_source
+WHERE {
```
I wonder if there are cases of xrefs that don't have any axioms, i.e. things matched by

https://github.com/related-sciences/nxontology-data/blob/c7b14299b18ed391762758bf76d8f5f806006a60/nxontology_data/efo/queries/xrefs.rq#L28-L29

We could do this match first and then make the axiom match OPTIONAL. Or we could decide this query is only for getting xref sources and we don't care about anything without a source.

Questions:

1. Do all `oboInOwl:hasDbXref` triples have corresponding axioms?
2. Do all axioms with `owl:annotatedProperty oboInOwl:hasDbXref` have corresponding `oboInOwl:hasDbXref` triples?

bfoltyn replied:
> I wonder if there are cases of xrefs that don't have any axioms, i.e. things matched by

There are cases like that, for example `MONDO:0004947` in `EFO:0000094` and `ICD10:O35` in `EFO:0009682` don't have axioms.



> We could do this match first and then make the axiom match OPTIONAL. Or we could decide this query is only for getting xref sources and we don't care about anything without a source.

I think that this query should be only for getting xref sources

dhimmel commented on nxontology_data/efo/queries/sources.rq:
Let's rename to `xref_sources.rq`

bfoltyn replied:
Ok

dhimmel commented on nxontology_data/efo/queries/mapping_properties.rq:
```diff
@@ -0,0 +1,28 @@
+  BIND( REPLACE( STR(?xref_uri), "^http.+/([^:]+)_(.+)$", "$1:$2" ) AS ?xref_id_dirty )
+  BIND(
+    COALESCE(
+        IF( STRSTARTS( ?xref_id_dirty, "http://identifiers.org" ), REPLACE( ?xref_id_dirty, "^http.*/(.+)/(.+)$", "$1:$2" ), ?error ),
+        IF( STRSTARTS( ?xref_id_dirty, "http://linkedlifedata.com/resource/umls/id" ), REPLACE( ?xref_id_dirty, "^http.*/(.+)$", "UMLS:$1" ), ?error ),
+        IF( STRSTARTS( ?xref_id_dirty, "http://purl.bioontology.org/ontology/ICD10CM" ), REPLACE( ?xref_id_dirty, "^http.*/(.+)$", "ICD10CM:$1" ), ?error ),
+        IF( STRSTARTS( ?xref_id_dirty, "https://icd.who.int/browse10/2019/en#" ), REPLACE( ?xref_id_dirty, "^http.*/(.+)$", "ICD10:$1" ), ?error ),
+        IF( STRSTARTS( ?xref_id_dirty, "https://omim.org/entry" ), REPLACE( ?xref_id_dirty, "^http.*/(.+)$", "OMIM:$1" ), ?error ),
+        IF( STRSTARTS( ?xref_id_dirty, "https://omim.org/phenotypicSeries" ), REPLACE( ?xref_id_dirty, "^http.*/PS(.+)$", "OMIMPS:$1" ), ?error ),
```
All these special cases are a bit annoying to maintain, but great work figuring them out. Was it just an iterative process of figuring out which URIs are not handled?

One option would be to save the URI to CURIE conversion for post-processing in python with [`bioregistry.curie_from_iri`](https://bioregistry.readthedocs.io/en/latest/api/bioregistry.curie_from_iri.html#bioregistry.curie_from_iri) or [`curies`](https://github.com/cthoyt/curies).

bfoltyn replied:
> Was it just an iterative process of figuring out which URIs are not handled?

Yes, this was an iterative process



> One option would be to save the URI to CURIE conversion for post-processing in python with [`bioregistry.curie_from_iri`](https://bioregistry.readthedocs.io/en/latest/api/bioregistry.curie_from_iri.html#bioregistry.curie_from_iri) or [`curies`](https://github.com/cthoyt/curies).

Thanks for the tip, this would be better if we can do it like that.  

I used `curies.get_bioregistry_converter` to get the converter for URIs. However, there are some differences how `curies` maps URIs compared to how I mapped them.  Here are some examples:
- `orphanet:99022` vs `obo:orphanet_99022` URI: `http://purl.obolibrary.org/obo/Orphanet_99022`.  Should we replace `obo:orphanet_` with `orphanet:`?
-  `orphanet:98813` vs `orphanet.ordo:98813` URI: `http://www.orpha.net/ORDO/Orphanet_98813` I guess we can replace `orphanet.ordo` with `orphanet` akin to https://github.com/related-sciences/nxontology-data/blob/c7b14299b18ed391762758bf76d8f5f806006a60/nxontology_data/utils.py#L84-L88
- `omimps:203655` vs `omim.ps:203655` URI: `https://omim.org/phenotypicSeries/PS203655` . Should we replace `omim.ps` with `omimps`?


There is also a missing uri_prefix `http://purl.bioontology.org/ontology/ICD10CM/` for `ICD10CM`.  The [add_prefix](https://curies.readthedocs.io/en/latest/api/curies.Converter.html#curies.Converter.add_prefix) method for adding prefixes lacks `merge` option in the version we use. Would it be safe to update the `curies` version?

dhimmel replied:
> * Should we replace `obo:orphanet_` with `orphanet:`?

Yes. It's nice that curies handles that.

For orphanet, we can replace `orphanet.ordo` after normalization.

`omim.ps` is the correct normalized prefix.

Ideal is you call `normalize_parsed_curie` on the output of `curies.get_bioregistry_converter` so we get consistently formatted CURIES everywhere.

dhimmel commented on nxontology_data/efo/queries/mapping_properties.rq:
> Once the queries are done, what would be the next steps? How this data should be represented in the output?

We can save the tables as output and/or we can include in the nxontology. We'll have to decide how we want to represent this information as node data in networkx. Doing so is tricky because we have to decide to what extent users will want access to rawer forms versus a more consolidated but opinionated format.

bfoltyn replied:
@dhimmel How about saving the table in the output for now, so that users can access this data? We could create a follow up issue to discuss how we can represent this information as node data in networkx.

dhimmel reviewed (commented):
Awesome work! I leave you mostly with hard questions and decisions (:

bfoltyn commented:
@dhimmel I've added saving the xref sources and mapping properties tables as output.  I've also removed the mapping from URI to CURIE in `mapping_properties.rq` and used `curies` instead.

bfoltyn commented:
@dhimmel I've added the curie normalization using `normalize_parsed_curie`

dhimmel commented:
Okay merged and exporting mesh in https://github.com/related-sciences/nxontology-data/actions/runs/6424734070!

Nice work navigating this @bfoltyn
